File: MPI/lab03/ex4/pi_paralelo_v2.c
/* ---------------------------------------------------------------
 * MPI pi Calculation Example - C Version 
 * Collective Communication example
 * FILE: pi.c
 * OTHER FILES: dboard.c
 *
 *   This version uses one MPI routine to collect results
 *
 * AUTHOR: Roslyn Leibensperger. Converted to MPI: George L. Gusciora 
 *   (1/25/95)
 * LAST REVISED: 06/07/96 Blaise Barney
 * --------------------------------------------------------------- */


#include <stdlib.h>
#include <stdio.h>
void srandom (unsigned seed);
double dboard (int darts);
#define DARTS 5000      /* number of throws at dartboard */
#define ROUNDS 10       /* number of times "darts" is iterated */
#define MASTER 0        /* task ID of master task */

int main(argc,argv)
int argc;
char *argv[];
{
double	homepi,         /* value of pi calculated by current task */
	pisum,	        /* sum of tasks' pi values */
	pi,	        /* average of pi after "darts" is thrown */
	avepi;	        /* average pi value for all iterations */
int	taskid,	        /* task ID - also used as seed number */
	numtasks,       /* number of tasks */
	rc,             /* return code */
	i;
MPI_Status status;

   /* Obtain number of tasks and task ID */




   if (rc != 0)
      printf ("error initializing MPI and obtaining task ID information\n");
   else
      printf ("task ID = %d\n", taskid);

   /* Set seed for random number generator equal to task ID */
   srandom (taskid);

   avepi = 0;
   for (i = 0; i < ROUNDS; i++)
   {
      /* All tasks calculate pi using dartboard algorithm */
      homepi = dboard(DARTS);
      /* Use MPI Routine across all tasks 
       */




      if (rc != 0)
         printf("%d: failure on mpc_reduce\n", taskid);

      /* Master computes average for this iteration and all iterations */
      if (taskid == MASTER)
      {
         pi = pisum/numtasks;
         avepi = ((avepi * i) + pi)/(i + 1); 
         printf("   After %3d throws, average value of pi = %10.8f\n",
                   (DARTS * (i + 1)),avepi);
      }    
   } 



   return 0;
}

File: lab03/ex4/pi_paralelo_v2.f
C  ------------------------------------------------------------------------
C  MPI pi Calculation Example - Fortran Version
C  Collective Communications examples
C  FILE: pi.f 
C  OTHER FILES: dboard.f
C
C  This version uses one MPI routine to collect results
C 
C  AUTHOR: Roslyn Leibensperger. Converted to MPI: George L. Gusciora (1/23/95)
C  LAST REVISED: 12/14/95 Blaise Barney
C  ------------------------------------------------------------------------
C Explanation of constants and variables used in this program:
C   DARTS          = number of throws at dartboard 
C   ROUNDS         = number of times "DARTS" is iterated 
C   MASTER         = task ID of master task
C   taskid         = task ID of current task 
C   numtasks       = number of tasks
C   homepi         = value of pi calculated by current task
C   pisum          = sum of tasks' pi values 
C   pi 	           = average of pi for this iteration
C   avepi          = average pi value for all iterations 
C   seednum        = seed number - based on taskid
C  ------------------------------------------------------------------------

      program pi_reduce


      integer DARTS, ROUNDS, MASTER
      parameter(DARTS = 5000) 
      parameter(ROUNDS = 10)
      parameter(MASTER = 0)

      integer 	taskid, numtasks, i, status(MPI_STATUS_SIZE)
      real*4	seednum
      real*8 	homepi, pi, avepi, pisum, dboard
      external  d_vadd

C     Obtain number of tasks and task ID



      write(*,*)'task ID = ', taskid

C     Use the task id to set the seed number for the random number generator.
      seednum = real(taskid)
      call srand(seednum)
      avepi = 0

      do 40 i = 1, ROUNDS
C     Calculate pi using dartboard algorithm 
        homepi = dboard(DARTS)

C     Use mpi routine  across all tasks




C     Master computes average for this iteration and all iterations 
        if (taskid .eq. MASTER) then
          pi = pisum/numtasks
          avepi = ((avepi*(i-1)) + pi) / i
          write(*,32) DARTS*i, avepi
 32       format('   After',i6,' throws, average value of pi = ',f10.8) 
        endif
 40   continue




      end
